lib/rubygems/package/tar_reader.rb

```diff
@@ -67,18 +69,22 @@ def each
       skip = (512 - (size % 512)) % 512
       pending = size - entry.bytes_read
 
-      begin
-        # avoid reading...
-        @io.seek pending, IO::SEEK_CUR
-        pending = 0
-      rescue Errno::EINVAL, NameError
-        while pending > 0 do
-          bytes_read = @io.read([pending, 4096].min).size
-          raise UnexpectedEOF if @io.eof?
-          pending -= bytes_read
+      if use_seek
+        begin
+          # avoid reading if the @io supports seeking
+          @io.seek pending, IO::SEEK_CUR
+          pending = 0
+        rescue Errno::EINVAL
         end
       end
 
+      # if seeking didn't work
```

Comment:
I guess "if seeking is not supported" is more adequate here, the current wording seems more appropriate for the old way of rescuing the `NameError`?

Reply:
Well, It could also be caused by the `rescue Errno::EINVAL`, but you're right, I'll clarify.

Reply:
Indeed, I missed that code path :+1: